myhpi/polls/models.py

```diff
@@ -98,26 +123,174 @@ def num_votes(self):
         return self.choices.aggregate(Sum("votes")).get("votes__sum")
 
 
-class Choice(models.Model):
+class MajorityVoteChoice(Orderable):
     text = models.CharField(max_length=254)
     votes = models.IntegerField(default=0)
+    page = ParentalKey("MajorityVotePoll", on_delete=models.CASCADE, related_name="choices")
 
     panels = [
         FieldPanel("text"),
     ]
 
-    class Meta:
-        abstract = True
-
     def __str__(self):
         return self.text
 
-
-class PollChoice(Orderable, Choice):
-    page = ParentalKey("polls.Poll", on_delete=models.CASCADE, related_name="choices")
-
     def percentage(self):
-        participant_count = self.page.participants.count()
+        participant_count = self.page.already_voted.count()
         if participant_count == 0:
             return 0
         return self.votes * 100 / participant_count
+
+
+class RankedChoicePoll(BasePoll):
+    content_panels = Page.content_panels + [
+        FieldPanel("description", classname="full"),
+        FieldPanel("start_date"),
+        FieldPanel("end_date"),
+        FieldPanel("eligible_groups"),
+        FieldPanel("results_visible"),
+        InlinePanel("options", label="Options"),
+    ]
+
+    parent_page_types = [
+        "PollList",
+    ]
+    subpage_types = []
+    search_fields = BasePage.search_fields + [
+        index.SearchField("description"),
+    ]
+
+    is_creatable = True
+
+    def cast_vote(self, request, *args, **kwargs):
+        form = self.get_ballot_form(request.POST)
+        if not form.is_valid():
+            messages.error(
+                request, mark_safe(_("Invalid ballot.\n{errors}").format(errors=form.errors))
+            )
+        else:
+            try:
```

Comment:
Can I theoretically send multiple valid form responses in the same moment and create a condition where I would be able to vote multiple times (since already_voted is only set at the end)?

Reply:
Hmm, I don't know much about Django ORM, but could we maybe move the uniqueness of votes constraint to the database, perform the insert of the user before the insert of the ballot and error out on a double insertion of the first? 

This way, we wouldn't have to worry about read-write locking the voter database.